# Market Basket Analyse/ Assoziationsregeln

Das Lernen von Assoziationsregeln (Association rule learning) ist ein Verfahren, um aus Datenmengen automatisiert Regeln oder Abhängigkeiten zu lernen.

Ein sehr bekanntes Beispiel hierfür ist die Anzeige von "Kunden, die dies kauften, kauften auch ...." auf Amazon. Das wird auch Market Basket Analysis genannt.

Wir schauen uns hier anhand eines beispielhaften Datensatzes eines Retail-Laden das im Detail an. Den Datensatz können Sie inkl. kurzer Erklärung [hier](https://archive.ics.uci.edu/dataset/352/online+retail) finden.

## Daten organisieren

In [1]:
# Importieren der Standard-Libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Load the file into pandas
retail_df = pd.read_excel("http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx")

## Datenaufbereitung

In [3]:
retail_df.describe(include="all")

InvoiceNo StockCode                         Description  \
count    541909.0    541909                              540455   
unique    25900.0      4070                                4223   
top      573585.0    85123A  WHITE HANGING HEART T-LIGHT HOLDER   
freq       1114.0      2313                                2369   
mean          NaN       NaN                                 NaN   
min           NaN       NaN                                 NaN   
25%           NaN       NaN                                 NaN   
50%           NaN       NaN                                 NaN   
75%           NaN       NaN                                 NaN   
max           NaN       NaN                                 NaN   
std           NaN       NaN                                 NaN   

             Quantity                    InvoiceDate      UnitPrice  \
count   541909.000000                         541909  541909.000000   
unique            NaN                            NaN            NaN   
top               NaN                            NaN            NaN   
freq              NaN                            NaN            NaN   
mean         9.552250  2011-07-04 13:34:57.156386048       4.611114   
min     -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%          1.000000            2011-03-28 11:34:00       1.250000   
50%          3.000000            2011-07-19 17:17:00       2.080000   
75%         10.000000            2011-10-19 11:27:00       4.130000   
max      80995.000000            2011-12-09 12:50:00   38970.000000   
std        218.081158                            NaN      96.759853   

           CustomerID         Country  
count   406829.000000          541909  
unique            NaN              38  
top               NaN  United Kingdom  
freq              NaN          495478  
mean     15287.690570             NaN  
min      12346.000000             NaN  
25%      13953.000000             NaN  
50%      15152.000000             NaN  
75%      16791.000000             NaN  
max      18287.000000             NaN  
std       1713.600303             NaN

In [4]:
retail_df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Wir sehen daß uns sehr viele Kundennummern fehlen. Deshalb löschen wir diese Spalte.

In [5]:
modified_retail_df = retail_df
modified_retail_df['Description'] = modified_retail_df['Description'].str.strip()

In der Dokumentation zum Datensatz steht daß alle Invoice-Nummern, die mit C anfangen, abgebrochene Bestellungen sind. Deshalb löschen wir die jetzt.

In [6]:
modified_retail_df['InvoiceNo'].str.contains('C').sum()

9288

Es gibt 9288 Bestellungen, die abgebrochen wurden

In [7]:
# Umwandeln der Spalte in Zeichenkette
modified_retail_df.dropna(axis = 0, subset =['InvoiceNo'], inplace = True)
modified_retail_df["InvoiceNo"] = modified_retail_df["InvoiceNo"].astype('str')

# damit wir nach Spalten mit C für cancelled filtern können
modified_retail_df = modified_retail_df[~modified_retail_df["InvoiceNo"].str.contains("C")]

In [8]:
modified_retail_df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1455
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     134697
Country             0
dtype: int64

Wir haben noch 1454 Zeilen ohne Artikelbeschreibung. Diese löschen wir; es hilft bei Regeln nicht viel, wenn wir den Artikelnamen nicht kennen.

In [9]:
modified_retail_df = modified_retail_df.dropna()

In [10]:
modified_retail_df.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [11]:
modified_retail_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899          CHILDREN'S APRON DOLLY GIRL         6   
541906    581587     23254         CHILDRENS CUTLERY DOLLY GIRL         4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138         BAKING SET 9 PIECE RETROSPOT         3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[397924 rows x 8 columns]

Wir sehen daß die ersten 4 Zeilen die gleiche Invoice-Nummer haben, also eine gleiche Bestellung sind.
Für eine Basketanalyse müssen wir jetzt noch eine Art virtuellen Einkaufskorb erstellen. Dazu verwenden wir den groupby-Befehl.

Dieser sagt, daß wir alle Zeilen, die die gleiche Invoice-Nummer haben, gruppieren in eine neue Zeile in einer Liste, gleiche Elemente aufaddieren.

Wir schauen uns nun an, wie sich Kaufaktionen in verschiedenen Ländern gestalten.

In [12]:
modified_retail_df['Country'].value_counts()

Country
United Kingdom          354345
Germany                   9042
France                    8342
EIRE                      7238
Spain                     2485
Netherlands               2363
Belgium                   2031
Switzerland               1842
Portugal                  1462
Australia                 1185
Norway                    1072
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     248
Unspecified                244
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58
Lebanon                     45
Lithuania                   35
Brazil                      32
Czech Republic              25
Bahrain                     17
Saudi Arabia                 9
Name: count, dtype: int64

UK und Deutschland sind die häufigsten, also schauen wir uns nur die beiden an.

In [13]:
# Befehl entnommen: https://www.geeksforgeeks.org/implementing-apriori-algorithm-in-python/

basket_UK = (modified_retail_df[modified_retail_df['Country'] =="United Kingdom"] # filtere auf alle Käufe die in England getan wurden
          .groupby(['InvoiceNo', 'Description'])['Quantity']  # gruppeiere anhand Bestellnummer und Beschreibung
          .sum()  # summiere die einzeln gefunden Elemente auf
          .unstack()
          .reset_index() # erstelle einen neuen Index
          .fillna(0)  # fülle leere Zeilen mit 0 auf
          .set_index('InvoiceNo'))  # Zeilenindex ist auf Invoice.Nr.

basket_GER = (modified_retail_df[modified_retail_df['Country'] =="Germany"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum()
          .unstack()
          .reset_index()
          .fillna(0)
          .set_index('InvoiceNo'))

basket_FRA = (modified_retail_df[modified_retail_df['Country'] =="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum()
          .unstack()
          .reset_index()
          .fillna(0)
          .set_index('InvoiceNo'))



Für den Assoziationsregelalgorithmus müssen wir noch eine Sache machen: der Algorithmus verlangt daß wir statt der Anzahl an gekauften Items einen False-Wert haben, falls nichts gekauft wurde und True sonst.

In [14]:
def encode_values(x):
    if(x <= 0):
        return 0
    if(x >= 1):
        return 1

# Encoding the datasets
basket_UK = basket_UK.applymap(encode_values)
basket_GER = basket_GER.applymap(encode_values)
basket_FRA = basket_FRA.applymap(encode_values)

/tmp/ipython-input-3347687889.py:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_UK = basket_UK.applymap(encode_values)
/tmp/ipython-input-3347687889.py:9: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_GER = basket_GER.applymap(encode_values)
/tmp/ipython-input-3347687889.py:10: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_FRA = basket_FRA.applymap(encode_values)


## Berechnen von Assoziationsregeln

Wir laden erst die notwendigen Bibliotheken

In [15]:
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import apriori, association_rules

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [16]:
# Model erstellen für Deutschland.
frq_items_GER = apriori(basket_GER,
                       min_support = 0.1, # mindestens 10% Support muss eine Regel haben
                       use_colnames = True)

# speichern der Regeln
rules_GER = association_rules(frq_items_GER,
                              metric ="lift",
                              min_threshold = 1)
# sortieren der Regeln
rules_GER = rules_GER.sort_values(['confidence', 'lift'],
                                  ascending =[False, False])

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [17]:
frq_items_GER

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

support                                           itemsets
0   0.102845                           (6 RIBBONS RUSTIC CHARM)
1   0.100656                       (JUMBO BAG WOODLAND ANIMALS)
2   0.115974                    (PLASTERS IN TIN CIRCUS PARADE)
3   0.107221                         (PLASTERS IN TIN SPACEBOY)
4   0.137856                 (PLASTERS IN TIN WOODLAND ANIMALS)
5   0.818381                                          (POSTAGE)
6   0.137856                         (REGENCY CAKESTAND 3 TIER)
7   0.157549                (ROUND SNACK BOXES SET OF 4 FRUITS)
8   0.245077               (ROUND SNACK BOXES SET OF4 WOODLAND)
9   0.102845                               (SPACEBOY LUNCH BOX)
10  0.126915                           (WOODLAND CHARLOTTE BAG)
11  0.100656           (PLASTERS IN TIN CIRCUS PARADE, POSTAGE)
12  0.100656                (PLASTERS IN TIN SPACEBOY, POSTAGE)
13  0.118162        (PLASTERS IN TIN WOODLAND ANIMALS, POSTAGE)
14  0.120350                (REGENCY CAKESTAND 3 TIER, POSTAGE)
15  0.150985       (ROUND SNACK BOXES SET OF 4 FRUITS, POSTAGE)
16  0.225383      (ROUND SNACK BOXES SET OF4 WOODLAND, POSTAGE)
17  0.115974                  (WOODLAND CHARLOTTE BAG, POSTAGE)
18  0.131291  (ROUND SNACK BOXES SET OF 4 FRUITS, ROUND SNAC...
19  0.124726  (ROUND SNACK BOXES SET OF 4 FRUITS, ROUND SNAC...

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [18]:
rules_GER

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents  \
8                 (ROUND SNACK BOXES SET OF 4 FRUITS)   
16  (ROUND SNACK BOXES SET OF 4 FRUITS, ROUND SNAC...   
2                          (PLASTERS IN TIN SPACEBOY)   
10               (ROUND SNACK BOXES SET OF4 WOODLAND)   
12                           (WOODLAND CHARLOTTE BAG)   
6                          (REGENCY CAKESTAND 3 TIER)   
0                     (PLASTERS IN TIN CIRCUS PARADE)   
4                  (PLASTERS IN TIN WOODLAND ANIMALS)   
14                (ROUND SNACK BOXES SET OF 4 FRUITS)   
17       (ROUND SNACK BOXES SET OF 4 FRUITS, POSTAGE)   
19                (ROUND SNACK BOXES SET OF 4 FRUITS)   
18      (ROUND SNACK BOXES SET OF4 WOODLAND, POSTAGE)   
15               (ROUND SNACK BOXES SET OF4 WOODLAND)   
20               (ROUND SNACK BOXES SET OF4 WOODLAND)   
11                                          (POSTAGE)   
9                                           (POSTAGE)   
21                                          (POSTAGE)   
7                                           (POSTAGE)   
5                                           (POSTAGE)   
13                                          (POSTAGE)   
3                                           (POSTAGE)   
1                                           (POSTAGE)   

                                          consequents  antecedent support  \
8                                           (POSTAGE)            0.157549   
16                                          (POSTAGE)            0.131291   
2                                           (POSTAGE)            0.107221   
10                                          (POSTAGE)            0.245077   
12                                          (POSTAGE)            0.126915   
6                                           (POSTAGE)            0.137856   
0                                           (POSTAGE)            0.115974   
4                                           (POSTAGE)            0.137856   
14               (ROUND SNACK BOXES SET OF4 WOODLAND)            0.157549   
17               (ROUND SNACK BOXES SET OF4 WOODLAND)            0.150985   
19      (ROUND SNACK BOXES SET OF4 WOODLAND, POSTAGE)            0.157549   
18                (ROUND SNACK BOXES SET OF 4 FRUITS)            0.225383   
15                (ROUND SNACK BOXES SET OF 4 FRUITS)            0.245077   
20       (ROUND SNACK BOXES SET OF 4 FRUITS, POSTAGE)            0.245077   
11               (ROUND SNACK BOXES SET OF4 WOODLAND)            0.818381   
9                 (ROUND SNACK BOXES SET OF 4 FRUITS)            0.818381   
21  (ROUND SNACK BOXES SET OF 4 FRUITS, ROUND SNAC...            0.818381   
7                          (REGENCY CAKESTAND 3 TIER)            0.818381   
5                  (PLASTERS IN TIN WOODLAND ANIMALS)            0.818381   
13                           (WOODLAND CHARLOTTE BAG)            0.818381   
3                          (PLASTERS IN TIN SPACEBOY)            0.818381   
1                     (PLASTERS IN TIN CIRCUS PARADE)            0.818381   

    consequent support   support  confidence      lift  representativity  \
8             0.818381  0.150985    0.958333  1.171012               1.0   
16            0.818381  0.124726    0.950000  1.160829               1.0   
2             0.818381  0.100656    0.938776  1.147113               1.0   
10            0.818381  0.225383    0.919643  1.123735               1.0   
12            0.818381  0.115974    0.913793  1.116587               1.0   
6             0.818381  0.120350    0.873016  1.066760               1.0   
0             0.818381  0.100656    0.867925  1.060539               1.0   
4             0.818381  0.118162    0.857143  1.047364               1.0   
14            0.245077  0.131291    0.833333  3.400298               1.0   
17            0.245077  0.124726    0.826087  3.370730               1.0   
19            0.225383  0.124726    0.791667  3.512540               1.0   
18            0.157549  0.124726    0.553398  

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [19]:
frq_items = apriori(basket_UK, min_support = 0.1, use_colnames = True)
rules_UK = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules_UK = rules_UK.sort_values(['confidence', 'lift'], ascending =[False, False])

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [20]:
rules_UK;

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [21]:
frq_items = apriori(basket_FRA, min_support = 0.1, use_colnames = True)
rules_FRA = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules_FRA = rules_FRA.sort_values(['confidence', 'lift'], ascending =[False, False])

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [22]:
rules_FRA

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents  \
49  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
48  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
41                    (SET/6 RED SPOTTY PAPER PLATES)   
44           (SET/6 RED SPOTTY PAPER PLATES, POSTAGE)   
34                (STRAWBERRY LUNCH BOX WITH CUTLERY)   
26               (ROUND SNACK BOXES SET OF4 WOODLAND)   
18                               (RABBIT NIGHT LIGHT)   
40                      (SET/6 RED SPOTTY PAPER CUPS)   
22                    (RED TOADSTOOL LED NIGHT LIGHT)   
10                    (PLASTERS IN TIN CIRCUS PARADE)   
43             (SET/6 RED SPOTTY PAPER CUPS, POSTAGE)   
4                                (LUNCH BAG WOODLAND)   
30                      (SET/6 RED SPOTTY PAPER CUPS)   
14                         (PLASTERS IN TIN SPACEBOY)   
32                    (SET/6 RED SPOTTY PAPER PLATES)   
0                            (LUNCH BAG APPLE DESIGN)   
24                         (REGENCY CAKESTAND 3 TIER)   
42  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...   
20                         (RED RETROSPOT MINI CASES)   
28               (SET/20 RED RETROSPOT PAPER NAPKINS)   
50  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...   
16                 (PLASTERS IN TIN WOODLAND ANIMALS)   
6                  (LUNCH BOX WITH CUTLERY RETROSPOT)   
46                    (SET/6 RED SPOTTY PAPER PLATES)   
39                    (SET/6 RED SPOTTY PAPER PLATES)   
2                           (LUNCH BAG RED RETROSPOT)   
52                    (SET/6 RED SPOTTY PAPER PLATES)   
12                         (PLASTERS IN TIN SPACEBOY)   
38               (SET/20 RED RETROSPOT PAPER NAPKINS)   
36               (SET/20 RED RETROSPOT PAPER NAPKINS)   
51               (SET/20 RED RETROSPOT PAPER NAPKINS)   
45                      (SET/6 RED SPOTTY PAPER CUPS)   
37                      (SET/6 RED SPOTTY PAPER CUPS)   
53                      (SET/6 RED SPOTTY PAPER CUPS)   
13                 (PLASTERS IN TIN WOODLAND ANIMALS)   
8                     (PLASTERS IN TIN CIRCUS PARADE)   
9                  (PLASTERS IN TIN WOODLAND ANIMALS)   
19                                          (POSTAGE)   
23                                          (POSTAGE)   
27                                          (POSTAGE)   
11                                          (POSTAGE)   
17                                          (POSTAGE)   
3                                           (POSTAGE)   
31                                          (POSTAGE)   
35                                          (POSTAGE)   
15                                          (POSTAGE)   
21                                          (POSTAGE)   
7                                           (POSTAGE)   
29                                          (POSTAGE)   
33                                          (POSTAGE)   
1                                           (POSTAGE)   
25                                          (POSTAGE)   
5                                           (POSTAGE)   
47                                          (POSTAGE)   

                                          consequents  antecedent support  \
49                    (SET/6 RED SPOTTY PAPER PLATES)            0.102828   
48                      (SET/6 RED SPOTTY PAPER CUPS)            0.102828   
41                      (SET/6 RED SPOTTY PAPER CUPS)            0.128535   
44                      (SET/6 RED SPOTTY PAPER CUPS)            0.107969   
34                                          (POSTAGE)            0.123393   
26                                          (POSTAGE)            0.159383   
18                                          (POSTAGE)            0.187661   
40                    (SET/6 RED SPOTTY PAPER PLATES)            0.138817   
22                                          (POSTAGE)            0.179949   
10                                          (POSTAGE)            0.169666   
43                    (SET/6 RED SPOTTY PAPER PLATES)      

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag